File: BOJ/1439 뒤집기.js
// 1439 뒤집기
// https://www.acmicpc.net/problem/1439

// sol.1) 그리디
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : 'text.txt')
    .toString().trim();

let zero = 0, one = 0;
let prev = '\0';
for (const c of input)
    if (prev != c) {
        c == '0' ? zero++ : one++;
        prev = c;
    }

console.log(Math.min(zero, one));


// sol.2) 정규 표현식
/*
    "000"과 같이 1이 존재하지 않을 수 있으며, 이떄 match().length를 사용할 시 오류가 발생.
    이를 예방하기 위해 match()의 반환값이 존재하지 않을 시 빈 배열 []을 넣어 length를 0으로 만들어줌.
    참고: https://www.acmicpc.net/source/38470467
*/
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : 'text.txt')
    .toString().trim();

const zero = (input.match(/0+/g) || []).length;
const one = (input.match(/1+/g) || []).length;

console.log(Math.min(zero, one));

File: BOJ/1439 뒤집기.cpp
// 1439 뒤집기
// https://www.acmicpc.net/problem/1439
/*
    그리디 (2,020KB, 0ms)
    시간 복잡도: O(n)
    풀이)
    - 0 또는 1이 연속되어 존재하는 횟수를 각각 세어준 후, 더 작은 값을 출력.
    - '0' = 48, '1' = 49
 */

#include <iostream>

using namespace std;
#define FAST_IO ios::sync_with_stdio(false), cin.tie(NULL), cout.tie(NULL);

int main()
{
    FAST_IO;

    string s;
    cin >> s;

    int v[51];
    v['0'] = v['1'] = 0;

    char prev = '\0';
    for (const auto &c : s)
        if (prev != c) // 이전 값과 다름. 즉, 새로운 값
            prev = c, v[c]++;

    cout << min(v['0'], v['1']);
}